domain/extract.go

```diff
@@ -0,0 +1,351 @@
+	"math/rand"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/BurntSushi/toml"
+	"github.com/pingcap/errors"
+	"github.com/pingcap/tidb/config"
+	"github.com/pingcap/tidb/kv"
+	"github.com/pingcap/tidb/parser/model"
+	"github.com/pingcap/tidb/sessionctx"
+	"github.com/pingcap/tidb/types"
+	"github.com/pingcap/tidb/util/logutil"
+	"github.com/pingcap/tidb/util/sqlexec"
+	"go.uber.org/zap"
+)
+
+const (
+	// ExtractMetaFile indicates meta file for extract
+	ExtractMetaFile = "meta.txt"
+)
+
+const (
+	// ExtractTaskType indicates type of extract task
+	ExtractTaskType = "taskType"
+)
+
+// ExtractType indicates type
+type ExtractType uint8
+
+const (
+	// ExtractPlanType indicates extract plan task
+	ExtractPlanType ExtractType = iota
+)
+
+func taskTypeToString(t ExtractType) string {
+	switch t {
+	case ExtractPlanType:
+		return "Plan"
+	}
+	return "Unknown"
+}
+
+// extractHandle handles the extractWorker to run extract the information task like Plan or any others.
+// extractHandle will provide 2 mode for extractWorker:
+// 1. submit a background extract task, the response will be returned after the task is started to be solved
+// 2. submit a task and wait until the task is solved, the result will be returned to the response.
+type extractHandle struct {
+	worker *extractWorker
+}
+
+// NewExtractHandler new extract handler
+func NewExtractHandler(sctxs []sessionctx.Context) *extractHandle {
+	h := &extractHandle{}
+	h.worker = newExtractWorker(sctxs[0], false)
+	return h
+}
+
+// ExtractTask extract tasks
+func (h *extractHandle) ExtractTask(ctx context.Context, task *ExtractTask, isBackgroundTask bool) (string, error) {
+	// TODO: support background job later
+	if isBackgroundTask {
+		return "", nil
+	}
+	return h.worker.extractTask(ctx, task)
+}
+
+type extractWorker struct {
+	ctx                context.Context
+	sctx               sessionctx.Context
+	isBackgroundWorker bool
+}
+
+// ExtractTask indicates task
+type ExtractTask struct {
+	extractType ExtractType
+
+	// variables for plan task type
+	Begin time.Time
+	End   time.Time
+}
+
+// NewExtractPlanTask returns extract plan task
+func NewExtractPlanTask(begin, end time.Time) *ExtractTask {
+	return &ExtractTask{
+		Begin:       begin,
+		End:         end,
+		extractType: ExtractPlanType,
+	}
+}
+
+func newExtractWorker(sctx sessionctx.Context, isBackgroundWorker bool) *extractWorker {
+	return &extractWorker{
+		sctx:               sctx,
+		isBackgroundWorker: isBackgroundWorker,
+	}
+}
+
+func (w *extractWorker) extractTask(ctx context.Context, task *ExtractTask) (string, error) {
+	switch task.extractType {
+	case ExtractPlanType:
+		return w.extractPlanTask(ctx, task)
+	}
+	return "", nil
+}
+
+func (w *extractWorker) extractPlanTask(ctx context.Context, task *ExtractTask) (string, error) {
+	if !config.GetGlobalConfig().Instance.StmtSummaryEnablePersistent {
+		return "", errors.New("tidb_stmt_summary_enable_persistent should be enabled for extract task")
```

Comment:
It seems like we cannot extract any information if tidb_stmt_summary_enable_persistent is disabled. Can we dump information from the memory `stmt_summary` table and generate a warning to notify the user the dumped information may not be completed in this case?